src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ReadWrite/ByteTests.cs

```diff
@@ -119,6 +119,19 @@ public void ReadByte_StructWithReferenceTypes_ReturnsExpected()
             };
 
             Assert.Equal(100, Marshal.ReadByte(structure, pointerOffset));
+
+            // The ReadByte() for object types does an explicit marshal which requires
+            // an allocation on each read. It can occur that the allocation is aligned
+            // on a byte boundary which would yield a value of 0. To mitigate the chance,
+            // marshal several times.
+            int readBytes = 0;
+            for (int i = 0; i < 3; ++i)
+            {
+                readBytes += Marshal.ReadByte(structure, stringOffset);
+            }
```

Comment:
> The offset isn't the issue

I'm saying, why can't we come up with a different offset to something that's not a pointer, an offset that points to one of the other fields that has a deterministic value?

Reply:
Ah, I see, we actually already do, both above and below this but cut off by the GitHub view of the diff.

Reply:
In that case, why do we believe looping three times is sufficient?  Why not instead read all of the bytes of the pointer using ReadByte, one by one, rather than repeatedly reading the same re-marshaled thing over and over and hoping for a better outcome?

Reply:
>Why not instead read all of the bytes of the pointer using ReadByte, one by one, rather than repeatedly reading the same re-marshaled thing over and over and hoping for a better outcome?

Because - and this is an implementation detail - each called to the `Read*()` APIs that have the `object` argument perform a distinct marshal of the type. Which means I would need to read the entire pointer first and then do as you suggest, which is a valid option. The alternative is to say there is very low probability that 3 calls will each yield an allocation that is aligned to the size of the primitive being read.

Reply:
> There's a, what, 1/8 chance each time?

I would need to think about the probabilities here, but that seems very high to me. I am unclear we can properly predict the probabilities without some discussion of the underlying allocation algorithm. I ran this locally and was observing the allocations - 1 out of 50 allocations were aligned in a way that would cause the issue. Clearly not very precise, but I would be very shocked if this hit that issue. the other test (16-bit) is much less likely to occur, so I am very confident on that stability of that test.

As far as deleting the test itself, I think this has value because the offset is to an object and not inligned to the value type as the other two are. If this fails at all, we can remove it, but would like to give it at least a day to see if this happens as often as feared.

Reply:
Even at 1/50, with 3 tries, that's 1/125K, and if the test is running, say, 2K times a day, it'll fail once every couple months.  For one test, that's maybe ok, but as we start to build up more and more tests that randomly fail with such frequency, it becomes a real productivity issue.  That's why I'm pushing on this: the only way to end up with good developer productivity over time is to drive such flakiness to zero, as otherwise enough of these "but it only fails once a month" failures creep in that it becomes a real problem.

If you really feel like this part of the test is valuable, then at least bump the 3 to something much higher so that the 1/50^N is small enough that's it's practically == never :)

Reply:
I will increase the count to 20.